util/rational/src/lib.rs

```diff
@@ -84,36 +97,60 @@ impl RationalU256 {
     }
 }
 
-// a/b * c/d = (a*c)/(b*d)
+// a/b * c/d = (a/gcd_ad)*(c/gcd_bc) / ((d/gcd_ad)*(b/gcd_bc))
 impl Mul<&RationalU256> for &RationalU256 {
     type Output = RationalU256;
     #[inline]
     fn mul(self, rhs: &RationalU256) -> RationalU256 {
-        RationalU256::new(&self.numer * &rhs.numer, &self.denom * &rhs.denom)
+        let gcd_ad = self.numer.gcd(&rhs.denom);
+        let gcd_bc = self.denom.gcd(&rhs.numer);
+
+        RationalU256::new_raw(
+            (&self.numer / &gcd_ad) * (&rhs.numer / &gcd_bc),
+            (&self.denom / gcd_bc) * (&rhs.denom / gcd_ad),
+        )
     }
 }
 
 impl Mul<RationalU256> for &RationalU256 {
     type Output = RationalU256;
     #[inline]
     fn mul(self, rhs: RationalU256) -> RationalU256 {
-        RationalU256::new(&self.numer * rhs.numer, &self.denom * rhs.denom)
+        let gcd_ad = self.numer.gcd(&rhs.denom);
+        let gcd_bc = self.denom.gcd(&rhs.numer);
+
+        RationalU256::new_raw(
+            (&self.numer / &gcd_ad) * (&rhs.numer / &gcd_bc),
+            (&self.denom / gcd_bc) * (&rhs.denom / gcd_ad),
+        )
```

Comment:
We can reuse the implementation of `&RationalU256 * &RationalU256`. Then it can be written like:

```
RationalU256 * &RationalU256 => (&self) * rhs
&RationalU256 * RationalU256 => self * (&rhs)
...
```

This code style saves many duplicated lines. But I don't know whether it is a good practice.

Reply:
update, please review again.